libstorage/fs.c

```diff
@@ -22,145 +22,157 @@ void storage_fsHandler(void *data, msg_t *msg)
 		case mtRead:
 			if (fs->ops->read == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.io.err = fs->ops->read(fs->info, &msg->i.io.oid, msg->i.io.offs, msg->o.data, msg->o.size);
+			msg->o.err = fs->ops->read(fs->info, &msg->oid, msg->i.io.offs, msg->o.data, msg->o.size);
 			break;
 
 		case mtWrite:
 			if (fs->ops->write == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.io.err = fs->ops->write(fs->info, &msg->i.io.oid, msg->i.io.offs, msg->i.data, msg->i.size);
+			msg->o.err = fs->ops->write(fs->info, &msg->oid, msg->i.io.offs, msg->i.data, msg->i.size);
 			break;
 
 		case mtTruncate:
 			if (fs->ops->truncate == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.io.err = fs->ops->truncate(fs->info, &msg->i.io.oid, msg->i.io.len);
+			msg->o.err = fs->ops->truncate(fs->info, &msg->oid, msg->i.io.len);
 			break;
 
 		case mtDevCtl:
 			if (fs->ops->devctl == NULL) {
 				/* FIXME this error passing works by accident on ioctl(),
 				 * there's no dedicated error field for devctl. */
-				msg->o.io.err = -ENOTTY; /* To return valid errno on ioctl() */
+				msg->o.err = -ENOTTY; /* To return valid errno on ioctl() */
 				break;
 			}
-			fs->ops->devctl(fs->info, &msg->i.io.oid, msg->i.raw, msg->o.raw);
+			fs->ops->devctl(fs->info, &msg->oid, msg->i.raw, msg->o.raw);
 			break;
 
 		case mtCreate:
 			if (fs->ops->create == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.create.err = fs->ops->create(fs->info, &msg->i.create.dir, msg->i.data, &msg->o.create.oid, msg->i.create.mode, msg->i.create.type, &msg->i.create.dev);
+			msg->o.err = fs->ops->create(fs->info, &msg->oid, msg->i.data, &msg->o.create.oid, msg->i.create.mode, msg->i.create.type, &msg->i.create.dev);
 			break;
 
 		case mtDestroy:
 			if (fs->ops->destroy == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.io.err = fs->ops->destroy(fs->info, &msg->i.destroy.oid);
+			msg->o.err = fs->ops->destroy(fs->info, &msg->oid);
 			break;
 
 		case mtSetAttr:
 			if (fs->ops->setattr == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.attr.err = fs->ops->setattr(fs->info, &msg->i.attr.oid, msg->i.attr.type, msg->i.attr.val, msg->i.data, msg->i.size);
+			msg->o.err = fs->ops->setattr(fs->info, &msg->oid, msg->i.attr.type, msg->i.attr.val, msg->i.data, msg->i.size);
 			break;
 
 		case mtGetAttr:
 			if (fs->ops->getattr == NULL) {
-				msg->o.io.err = -ENOSYS;
+				msg->o.err = -ENOSYS;
 				break;
 			}
-			msg->o.attr.err = fs->ops->getattr(fs->info, &msg->i.attr.oid, msg->i.attr.type, &msg->o.attr.val);
+			msg->o.err = fs->ops->getattr(fs->info, &msg->oid, msg->i.attr.type, &msg->o.attr.val);
+			break;
+
+		case mtGetAttrAll:
+			if (fs->ops->getattrall == NULL) {
+				msg->o.err = -ENOSYS;
+				break;
+			}
```

Comment:
Just an idea - maybe we could provide stop-gap implementation by iterating using `fs->ops->getattr` instead?

Reply:
Hmm, in theory it's a good idea but I don't really think there's a use case for it - I provided implementation for all filesystems that support `getattr` and if we were to add a support for a new one it's not much work to implement both `getattr` and `getattrall`.

Reply:
You're right, nice feature but dead code in practice